internal/engine/upper_test.go

```diff
@@ -61,6 +62,7 @@ func (b *fakeBuildAndDeployer) BuildAndDeploy(ctx context.Context, service model
 }
 
 func (b *fakeBuildAndDeployer) GetContainerForBuild(ctx context.Context, build BuildResult) (k8s.ContainerID, error) {
+	b.getContainerForBuildCount++
 	if build.Image == nil {
```

Comment:
do you understand why this test doesn't fail?

Reply:
Yeah because this is a test that the build fails. The build itself should still fail. It just shouldn't try to get the container for a build without an image.

 I don't understand the interaction between `nextBuildFailure` and `endToken` though. I can't 100% explain what this test is doing.